std/typecons.d

```diff
@@ -1175,7 +1175,12 @@ if (distinctFieldNames!(Specs))
             size_t h = 0;
             static foreach (i, T; Types)
             {{
-                const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
+                static if (__traits(compiles, h = .hashOf(field[i])))
+                    const k = .hashOf(field[i]);
+                else
+                    // Workaround for when .hashOf is not both @safe and nothrow.
+                    // BUG: Improperly casts away `shared`!
+                    const k = typeid(T).getHash((() @trusted => cast(const void*) &field[i])());
```

Comment:
Any reason not to add some `static if`s that check the type qualifiers and cast to the correct pointer type?

Reply:
That doesn't work :-/. If it did I wouldn't need to cast at all, I could just write ` &field[i]`.

Example:
https://run.dlang.io/is/ov8L8P
```d
void main()
{
    shared int a;
    int b = typeid(a).getHash(&a); // can't TypeInfo.getHash a shared argumennt
}
```

Reply:
To be clear, I don't think we should be doing this type of casting in `toHash`, but this was already being done in `Tuple` & `Unique` & `Typedef` so removing it could cause previously-compiling code using those templates to fail.

Reply:
Got it. Thanks for clarifying it for me.